[WebGPU] Hook up UnrealizedCoreTextFont to callers

litherum opened the pull request:
#### 3563129ef78f532b8395346823cd2d4f186fd499
<pre>
[WebGPU] Hook up UnrealizedCoreTextFont to callers
<a href="https://bugs.webkit.org/show_bug.cgi?id=252221">https://bugs.webkit.org/show_bug.cgi?id=252221</a>

Reviewed by Cameron McCormack.

The solution to <a href="https://bugs.webkit.org/show_bug.cgi?id=247987">https://bugs.webkit.org/show_bug.cgi?id=247987</a> requires having explicit locations where we transform
a CTFontDescriptor into a CTFont. UnrealizedCoreTextFont is the critical object in a design which enforces this: It
holds the original CTFontDescriptor/CTFont which Core Text gave us originally, plus a CFMutableDictionary of
modifications to be made to it. Updating the modifications never hits Core Text, and are as such very lightweight.
Once we&apos;re done setting the modifications, there&apos;s a &quot;commit&quot; step where we finally make the final CTFont. This
means that we avoid ping-ponging between CTFontDescriptors and CTFonts unnecessarily.

Eventually, preparePlatformFont() will be integrated into this object, but that&apos;s too big for a single patch. This
patch just hooks up callers of preparePlatformFont() to use this object, which will temporarily be resolved to a
CTFont just before calling preparePlatformFont. The next few patches will be updating preparePlatformFont() to be
able to represent it as a pure modification to the UnrealizedCoreTextFont.

* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::preparePlatformFont):
(WebCore::fontDescriptorWithFamilySpecialCase):
(WebCore::fontWithFamily):
(WebCore::FontCache::systemFallbackForCharacters):
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.h:
* Source/WebCore/platform/graphics/cocoa/FontFamilySpecificationCoreText.cpp:
(WebCore::FontFamilySpecificationCoreText::fontRanges const):
* Source/WebCore/platform/graphics/mac/FontCustomPlatformDataMac.cpp:
(WebCore::FontCustomPlatformData::fontPlatformData):

Canonical link: <a href="https://commits.webkit.org/260241@main">https://commits.webkit.org/260241@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ea39c78caca731a8eba2235ffea0bc34bb0f34f1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107655 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16707 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40551 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116803 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/116208 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111547 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18129 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8025 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99828 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113409 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/13685 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/96884 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/41357 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/95601 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28515 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83201 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9693 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29864 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10368 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/6756 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15844 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49449 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7078 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11928 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
```diff
@@ -981,38 +982,50 @@ static RetainPtr<CTFontDescriptorRef> fontDescriptorWithFamilySpecialCase(const
-        return descriptor;
+        if (traits) {
+            // FIXME: rdar://105369379 As far as I can tell, there's no modification to the attributes dictionary that has the same effect as CTFontDescriptorCreateCopyWithSymbolicTraits(),
+            // because there doesn't seem to be a place to specify the bitmask.
```
Can you mention in the comment that it's for that reason we create the CTFont eagerly?

heycam commented on Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
```diff
@@ -981,38 +982,50 @@ static RetainPtr<CTFontDescriptorRef> fontDescriptorWithFamilySpecialCase(const
+            CFTypeRef keys[] = { kCTFontFeatureTypeIdentifierKey, kCTFontFeatureSelectorIdentifierKey };
+            CFTypeRef values[] = { numberSpacingNumber.get(), monospacedNumbersNumber.get() };
+            ASSERT(std::size(keys) == std::size(values));
+            auto settingsDictionary = adoptCF(CFDictionaryCreate(kCFAllocatorDefault, keys, values, std::size(keys), &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks));
+            CFTypeRef entries[] = { settingsDictionary.get() };
+            auto settingsArray = adoptCF(CFArrayCreate(kCFAllocatorDefault, entries, std::size(entries), &kCFTypeArrayCallBacks));
+            CFDictionaryAddValue(attributes, kCTFontFeatureSettingsAttribute, settingsArray.get());
```
Is this a behavior change, or is that something already handled by the old CTFontDescriptorCreateCopyWithFeature call? If it's a change, it's worth calling out in the commit message (and maybe should've been a separate patch).

litherum replied:
It's not a behavior change 👍

heycam commented on Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
```diff
@@ -1022,12 +1035,17 @@ static RetainPtr<CTFontRef> fontWithFamily(FontDatabase& fontDatabase, const Ato
-        return preparePlatformFont(font.get(), fontDescription, fontCreationContext, true);
+    if (auto unrealizedFont = fontDescriptorWithFamilySpecialCase(family, fontDescription, size, fontDescription.shouldAllowUserInstalledFonts())) {
+        unrealizedFont->setSize(size);
+        unrealizedFont->modify([allowUserInstalledFonts = fontDescription.shouldAllowUserInstalledFonts()](CFMutableDictionaryRef attributes) {
```
I find it a bit weird to use a captured variable to store the result of this expression; WDYT about capturing the `fontDescription` reference, and calling `shouldAllowUserInstalledFonts()` inside the lambda body? Or just capture `[&]` which I think is very common in the codebase.

heycam commented on Source/WebCore/platform/graphics/mac/FontCustomPlatformDataMac.cpp:
```diff
@@ -40,17 +41,17 @@ FontCustomPlatformData::~FontCustomPlatformData() = default;
-    ASSERT(modifiedFontDescriptor);
+    UnrealizedCoreTextFont unrealizedFont = { RetainPtr { fontDescriptor } };
+    unrealizedFont.setSize(fontDescription.computedPixelSize());
+    unrealizedFont.modify([shouldAllowUserInstalledFonts = fontDescription.shouldAllowUserInstalledFonts()](CFMutableDictionaryRef attributes) {
```
And here.

litherum committed:
[WebGPU] Hook up UnrealizedCoreTextFont to callers
https://bugs.webkit.org/show_bug.cgi?id=252221

Reviewed by Cameron McCormack.

The solution to https://bugs.webkit.org/show_bug.cgi?id=247987 requires having explicit locations where we transform
a CTFontDescriptor into a CTFont. UnrealizedCoreTextFont is the critical object in a design which enforces this: It
holds the original CTFontDescriptor/CTFont which Core Text gave us originally, plus a CFMutableDictionary of
modifications to be made to it. Updating the modifications never hits Core Text, and are as such very lightweight.
Once we're done setting the modifications, there's a "commit" step where we finally make the final CTFont. This
means that we avoid ping-ponging between CTFontDescriptors and CTFonts unnecessarily.

Eventually, preparePlatformFont() will be integrated into this object, but that's too big for a single patch. This
patch just hooks up callers of preparePlatformFont() to use this object, which will temporarily be resolved to a
CTFont just before calling preparePlatformFont. The next few patches will be updating preparePlatformFont() to be
able to represent it as a pure modification to the UnrealizedCoreTextFont.

* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::preparePlatformFont):
(WebCore::fontDescriptorWithFamilySpecialCase):
(WebCore::fontWithFamily):
(WebCore::FontCache::systemFallbackForCharacters):
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.h:
* Source/WebCore/platform/graphics/cocoa/FontFamilySpecificationCoreText.cpp:
(WebCore::FontFamilySpecificationCoreText::fontRanges const):
* Source/WebCore/platform/graphics/mac/FontCustomPlatformDataMac.cpp:
(WebCore::FontCustomPlatformData::fontPlatformData):

Canonical link: https://commits.webkit.org/260241@main
